### DISCUSSION TITLE
Brazilian Libre 2 Plus - XDrip+ OOP2

### DISCUSSION THREAD

**Discussion by ibnbr** [NightscoutFoundation/xDrip#3545] [Ideas]
Is it possible for OOP2 to recognize the Brazilian Libre 2 Plus? I enabled the NFC log, and in the patch info, it shows 

![image](https://github.com/NightscoutFoundation/xDrip/assets/2778247/35062400-86cd-4b31-ac5d-15764a77cbdd)

**Comment by PhilTU** [comment]
case 0x76, 0x2B:
patchInfo[3] == 2 ? .libre2US :
patchInfo[3] == 4 ? .libre2CA :
patchInfo[3] == 8 ? .libre2RU :

seems same as Russia with 08

Brazil : 2B 0A 3A 08
Russia : 2B 0A 39 08

I think this is the type of sensor (2B) that requires different NFC commands which have not been implemented.

**Reply by ibnbr** [reply]
Any thing I can do/collect/test to help implement this SN#? I`m a developer (not Java), but maybe I can help.

**Comment by lua-vr** [comment]
I don't know how relevant this is, but Juggluco can read the brazilian Libre Plus: 
https://www.youtube.com/watch?v=-aEFtZHBLIo

Since it's open source (https://github.com/j-kaltes/Juggluco) perhaps xDrip can copy the relevant code?

**Reply by ibnbr** [reply]
Yeah I know, but juggluco dont sent historical data when lost connection, leaving the chart in xdrip/nightscout with a lot of gaps.

this day was with juggluco....
![image](https://github.com/NightscoutFoundation/xDrip/assets/2778247/38688e82-29ef-43e2-bc54-831179e472f3)


Thank you.

**Reply by lua-vr** [reply]
My suggestion was to make xDrip do the same as Juggluco to receive the sensor data. 

I don't have any Libre 2 Plus yet but did you try using the Libre 2/3 OOP that should work directly with xDrip? Since Juggluco works perhaps it could work too? I'm really curious.

**Reply by ibnbr** [reply]
XDRIP + OOP don't work with Brazillian Libre2 Plus, which is why I opened this discussion.

**Reply by PhilTU** [reply]
> Since it's open source

Sort of, it uses Librelink library / code too. Which is why you have to have Librelink installed or present as an apk.

I'm not sure this is acceptable to xDrip+ developers.

**Reply by taisatk** [reply]
> XDRIP + OOP don't work with Brazillian Libre2 Plus, which is why I opened this discussion.

Have you found out a workaround for this, @ibnbr ? I just started using FSL2+ (patchinfo 2B 0A 3A) and I'd really like to go on using xDrip.

**Reply by PhilTU** [reply]
> Have you found out a workaround

Juggluco app will connect to that sensor. You could just use that alone, or you can send data to xDrip+ see options at juggluco.nl you have choice of Eversense or Patched app inputs in xDrip

**Reply by ibnbr** [reply]
> > XDRIP + OOP don't work with Brazillian Libre2 Plus, which is why I opened this discussion.
> 
> Have you found out a workaround for this, @ibnbr ? I just started using FSL2+ (patchinfo 2B 0A 3A) and I'd really like to go on using xDrip.

nope. :-(

**Comment by juanpmartin** [comment]
Same in Argentina! Would be great to have that compatible.

**Comment by LTorrao** [comment]
@ibnbr @juanpmartin 
Hello/Olá/Hola from Portugal.
I have just now had the same issue because the Freestyle Libre 2 Plus is now being sold here too (replacing the Libre 2).
As I use Xdrip ecosystem with followers, it could be a big problem for me!
The solution I found was to set the source phone with xDrip as a nightscout follower.
Now, having Tomato installed, I can set it as a Nightscout source so, it is working, good news :)

**Reply by PhilTU** [reply]
Which version of OOP2 did you try?

**Reply by LTorrao** [reply]
I have it installed, but not running currently. I think it is not necessary in this case. My setup is
1. Main smartphone:
- Tomato connected to MiaoMiao 1, sending data to Phone via nightscout
- Main phone's xDrip app set as nightscout follower (getting the data from Tomato's nightscout)
2. Follower devices
- Simply follow in xDrip

Hope this helps :)

**Reply by PhilTU** [reply]
ok, not really interested in that but I was trying to determine if the Libre 2+ you had in Portugal didn't work directly with xDrip+.

you're reading the NFC with a transmitter device and using Tomato app to interpret the data so xDrip is just like a display only in this case.

**Reply by LTorrao** [reply]
Exactly. For me, the problem would be having my xDrip followers ecossistem down.

Here in PT, we did not have FL2 Plus up until a few days ago. I have been using FL2 for some years now, but in the farmacy, I was told this time that there were only Plus, but that it was the same as the 2, which I later realised is not. Today, when installing. I would see the data being loaded in the log on xDrip, but not a real number on screen. I tried an old OOP2 I had installed in the phone in case that helped, but it did not. This is when I saw this post about the Brazilian 2 Plus, and thought, maybe it is the same. Lucky for me, I had the nightscout site prepared from other setups, so it was just a matter of using Tomato to send the MiaoMiao1 data via nightscout and then make my main device xDrip a nightscout follower of that data. Then, my follower devices follow the main phone, as always.

I decided to share here because I see this problem is happening in the world. Brazil, Argentina, now Portugal. So I though, let me leave a comment explaining how I managed to sort it out, it may help someone :)

**Reply by PhilTU** [reply]
Unfortunately you haven't said what version of OOP2 (or xDrip) you tried to use. Libre 2+ in UK and EU works with both software from 31/5/24 or later.

I would expect Portugal is more likely to get EU than South America sensor types but I don't know.

**Reply by LTorrao** [reply]
Dear @PhilTU , thank you very much for this information, it might be very helpful for me if I have problems with MiaoMiao. But, please, could you clarify, to use Libre 2 Plus in EU and UK, I see I need to use Xdrip and OOP from 31/05/2024 or later. However, where can I find such versions? (do you have a link to the latest OOP, please?)

Regarding your question, I was using:
- xDrip dfcbe80-2024.09.17 Code: 2124091707
- OOP 27da3f5-2023.05.25 Code: 2123052502

**Reply by PhilTU** [reply]
https://xdrip.readthedocs.io/en/latest/use/OOP/?h=oop2

It says LINK HERE sometimes it is not so clear.

**Reply by LTorrao** [reply]
Thank you :)

**Reply by juanpmartin** [reply]
Unfortunately it's not working with sensors sold in argentina (S/N starting with 3MH0). I'm using juggluco in order to connect to the sensor.

PS: When I scan the sensor with xdrip it's selecting Libre 1 (failsafe).

**Reply by PhilTU** [reply]
> Unfortunately it's now working with sensors sold in argentina (S/N starting with 3MH0).

I think you mean "NOT" working with South American sensors. This is same with Australia/NZ and US / Canada.

**Reply by juanpmartin** [reply]
> > Unfortunately it's now working with sensors sold in argentina (S/N starting with 3MH0).
> 
> I think you mean "NOT" working with South American sensors. This is same with Australia/NZ and US / Canada.

Excactly (i've just edited it). Thanks!

**Comment by Logue-0** [comment]
## Brazilian Libre 2 Plus — same issue, traces available

Hi all. Same problem as the OP — Brazilian FreeStyle Libre 2 Plus sensors don't work with Juggluco, xDrip+ OOP2, or any patched LibreLink variant. Posting to confirm the issue, add more sensor data, and offer to contribute whatever raw data would help.

---

**Setup**
- Device: Samsung Galaxy S25 Ultra (Android 15)
- Sensor: FreeStyle Libre 2 Plus, Brazil
- Juggluco 10.9.2 — NFC reads succeed but algorithm rejects the data every time
- Abbott's official FreeStyle LibreLink app — eventually works via NFC (8h history) and BLE, but shows "activated on another device" after Juggluco's failed attempts corrupt the sensor session

---

**patchInfo bytes (from Juggluco 10.9.2 trace and logcat logs)**

```
2B 0A 3A 08 1F E1
```

- `patchInfo[0] = 0x2B` — hits the existing case branch in xDrip+'s sensor type switch
- `patchInfo[3] = 0x08` — currently mapped to `.libre2RU` in xDrip+'s logic, but this is a Brazilian sensor, not Russian. Worth investigating whether BR and RU share a region byte or whether this is a separate unrecognized variant.

`getProductFamily() = 3` (Gen 2 / libcalibrat2 algorithm family)

**Sensor NFC UIDs**
- May 2026 sensor: `E007A4001A40BF24`
- March 2026 sensor: `E007A4001851732C`

Sensor serial: `3MH01MH5Z4H`

---

**What Juggluco's algorithm returns on every scan attempt**

The following output is from Juggluco 10.9.2's trace.log (enabled via Left menu → Settings → Logging). Full trace.log and logcat files for both sensors are attached below.

March sensor (three fresh activation attempts):
```
errorCode=21
Is not inserted and not removed
processScan returned 4
Badscan 2
```

May sensor (after prior partial activation attempt):
```
Is inserted and removed
init starttime Wed May 20 19:58:54 2026
init endtime   Wed May 20 21:32:54 2026
Badscan 14
```
Followed by repeated Bluetooth GATT `onConnectionStateChange status=147` (GATT_CONN_FAIL_ESTABLISH) — BLE unreachable after failed activation.

---

**What works vs. what doesn't**

| Approach | Result |
|---|---|
| Abbott official LibreLink (BLE + NFC) | ✅ Works |
| LibreView cloud → nightscout-librelink-up bridge | ✅ Works (only third-party path) |
| Juggluco 10.9.2 | ❌ Algorithm rejects sensor at processScan |
| xDrip+ OOP2 | ❌ (confirmed by this thread) |
| Patched LibreLink (any variant) | ❌ No known version supports BR Libre 2 Plus |

The fact that Abbott's official app handles these sensors over BLE confirms the protocol is implementable — the command set just hasn't been reverse-engineered for this variant yet.

---

I have full Juggluco trace logs from both sensors (trace.log + logcat for each) including the complete NFC command/response sequences. Happy to attach them here or share however is most useful.

I'm not a Java developer but I can run tests, collect NFC logs, and provide payloads from fresh sensors if needed. What would be most helpful? These logs af
[NEW_trace.log](https://github.com/user-attachments/files/28129182/NEW_trace.log)
[NEW_logcat.txt](https://github.com/user-attachments/files/28129184/NEW_logcat.txt)
[OLD_logcat.txt](https://github.com/user-attachments/files/28129188/OLD_logcat.txt)
[OLD_trace.log](https://github.com/user-attachments/files/28129190/OLD_trace.log)

**Comment by Logue-0** [comment]
Hi, thx for the fast answer. So, I've posted this here instead of on juggluco's github because there the author claims his software is fine and my sensors are the one at fault, so I've given up on that.

Also, I believe that the library that juggluco uses for data decryption is really really OLD. I don't have confirmation on this, but it's the only thing that I could think of that breaks juggluco for BR sensors only while working everywhere else. And this was fairly recent, since up until March 2026 I used juggluco just fine. So whatever change happened was, for sure, after the start of 2026 (can't pin point the exact date due to app distribution policies, since I know Librelink doesn't release a version for everyone at the same time). So probably Abbott's implementation for this specific sensor variant and it's decryption was altered somehow and that is blocking juggluco from getting the proper data.

**Comment by mhalano** [comment]
I tested my sensor with Juggluco quite time ago, and did work. The problem is Juggluco interface is not ideal if compared with xDrip. I will test again when possible.

**Reply by mhalano** [reply]
I tested, actually I'm using right now, Juggluco 10.9.3 with Freestyle Libre 2 Plus from Brazil. It works very well the register. Unfortunately the app is hard to use, so I would like xDrip but reading the sensors as Juggluco. What I will try to do is sending the data to xDrip, so the Juggluco deals with the sensor nad xDrip just shows the data. Not great, but that what is available right now.
